Reduces redundancy by using only one function to handle all outgoing HTML mail

juho-jaakkola opened the pull request:
Also adds an attachment to the test.

Fixes #26
Fixes #27

juho-jaakkola commented:
I think this is good to go!

jeabakker commented:
the way you reused the function name html_email_handler_send_email() to be a plugin hook handler will break every plugin that uses that function (digest, newsletter to name a few).

this is bc-breaking ;) we don't want to do that.

Also why have an extra file for the attachment test, why not send for example the manifest

jeabakker commented on lib/hooks.php:
```diff
@@ -4,118 +4,25 @@
-	
-	return html_email_handler_send_email($options);
+function html_email_handler_email_hook($hook, $type, $options, $params) {
+	return html_email_handler_send_email($return);
```
$return is undefined should be $options

jeabakker commented on lib/functions.php:
```diff
@@ -157,9 +156,23 @@ function html_email_handler_send_email(array $options = null) {
+		$headers['Content-Type'] = "multipart/alternative; boundary=\"{$boundary}\"";
+	}
+	
+	global $CONFIG;
```
please don't use $CONFIG but use elgg_get_config() instead

juho-jaakkola commented on lib/functions.php:
```diff
@@ -80,6 +77,9 @@ function html_email_handler_send_email(array $options = null) {
 	}
 	
+	$options['html_message'] = html_email_handler_make_html_body($options);
+	$options['plaintext_message'] = $options['body'];
```
I guess these two should be added only if both of them are empty? Should I also check whether the global HTML email setting has been enabled?

jeabakker replied:
i think they should only be filled when not empty.

no need to check the plugin setting, this function is there to make sure you can send html mail regardless of the setting. Only the hooks need to respect the plugin setting

jeabakker commented:
please check the one tiny issue in scrutinizer, if fixed i can merge

juho-jaakkola commented:
done

jeabakker commented:
@Facyla could you maybe have a look at this pr. The attachements still work according to juho but since you build it you can see if it all works

Facyla commented:
Hi,

I'm sorry i will not have time today (and probably this whole week) to test it in detail, so a quick note after reading the changes : 

In hook, the "attachments" now get passed through $options, and not $params : as far as i read the discussion that was precisely what has been improved through this commit, but it modifies a little the way attachments are passed.
A further test would be checking that attachments also gets passed when using : notify_user() function with $params['attachments']
=> If this is OK, it's fine for me.

Sorry for not testing it myself...
Thank you,
Florian

jeabakker commented:
@Facyla no problem, thanks for the tip. 

I think you have a point maybe @juho-jaakkola can have a look at it

juho-jaakkola commented:
The ```$params``` should be used only for original core parameters (since 2012: 16cf5867f33bada9837158272a8e944347a4713b).

I suggest using it for now but giving a deprecation warning if the ```'attachments'``` option is found from there.

Facyla commented:
Ok for deprecation warning, that was mainly for bc reasons, but i totally agree that some behaviours require refactoring :)

juho-jaakkola commented:
> the "attachments" now get passed through $options, and not $params

This doesn't matter because Elgg core populates them both with the same values (e.g. the parameters passed into ```notify_user()```).

The ```notify_user()``` function does however pass the attachments under ```$options['params']```, so I added a fix that looks for them from ```$options['params']['attachments']``` instead of ```$options['attachments']```, so now they also work.

I also added a test that sends an email through ```notify_user()``` with an attachment.
